File: server/shared/logger.js
/**
 * Structured Logger — Winston-based logging for production
 * Provides consistent log format across both auth and agent servers.
 */
const winston = require('winston');

const { combine, timestamp, printf, colorize, errors } = winston.format;

// Custom log format
const logFormat = printf(({ level, message, timestamp, service, ...meta }) => {
    const metaStr = Object.keys(meta).length > 0
        ? ` ${JSON.stringify(meta)}`
        : '';
    return `${timestamp} [${service || 'app'}] ${level}: ${message}${metaStr}`;
});

/**
 * Create a logger instance for a service.
 * @param {string} service — 'auth' | 'agent'
 */
function createLogger(service = 'app') {
    const isProduction = process.env.NODE_ENV === 'production';

    const transports = [
        new winston.transports.Console({
            format: combine(
                colorize(),
                logFormat
            )
        })
    ];

    // In production, also log to files
    if (isProduction) {
        transports.push(
            new winston.transports.File({
                filename: `logs/${service}-error.log`,
                level: 'error',
                maxsize: 5 * 1024 * 1024, // 5MB
                maxFiles: 5
            }),
            new winston.transports.File({
                filename: `logs/${service}-combined.log`,
                maxsize: 10 * 1024 * 1024, // 10MB
                maxFiles: 5
            })
        );
    }

    return winston.createLogger({
        level: isProduction ? 'info' : 'debug',
        defaultMeta: { service },
        format: combine(
            errors({ stack: true }),
            timestamp({ format: 'YYYY-MM-DD HH:mm:ss' })
        ),
        transports,
        // Don't exit on uncaught exceptions
        exitOnError: false
    });
}

/**
 * Express request logging middleware.
 * Logs method, URL, status code, and response time.
 */
function requestLogger(logger) {
    return (req, res, next) => {
        const start = Date.now();

        // Log when response finishes
        res.on('finish', () => {
            const duration = Date.now() - start;
            const level = res.statusCode >= 500 ? 'error'
                : res.statusCode >= 400 ? 'warn'
                : 'info';

            logger.log(level, `${req.method} ${req.originalUrl} ${res.statusCode} ${duration}ms`, {
                method: req.method,
                url: req.originalUrl,
                status: res.statusCode,
                duration,
                ip: req.ip,
                userId: req.userId || req.user?._id || null
            });
        });

        next();
    };
}

/**
 * Global error handler middleware.
 * Catches unhandled errors and returns a standardized error response.
 */
function errorHandler(logger) {
    return (err, req, res, next) => {
        logger.error(`Unhandled error: ${err.message}`, {
            stack: err.stack,
            method: req.method,
            url: req.originalUrl,
            userId: req.userId || req.user?._id || null
        });

        // Don't leak error details in production
        const isProduction = process.env.NODE_ENV === 'production';
        res.status(err.status || 500).json({
            success: false,
            message: isProduction
                ? 'An internal server error occurred. Please try again.'
                : err.message,
            ...(isProduction ? {} : { stack: err.stack })
        });
    };
}

module.exports = { createLogger, requestLogger, errorHandler };

File: server_py/shared/logger.py
"""
Structured Logger — Loguru-based logging for production.
Provides consistent log format across auth and agent subsystems.
"""

import sys
import time
from pathlib import Path
from loguru import logger as _loguru_logger
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response

from config import get_settings


def _setup_logger():
    """Configure loguru with appropriate sinks."""
    settings = get_settings()

    # Remove default handler
    _loguru_logger.remove()

    # Console handler (always)
    log_format = (
        "<green>{time:YYYY-MM-DD HH:mm:ss}</green> "
        "[<cyan>{extra[service]}</cyan>] "
        "<level>{level: <8}</level> "
        "<level>{message}</level>"
    )
    _loguru_logger.add(
        sys.stderr,
        format=log_format,
        level="DEBUG" if not settings.is_production else "INFO",
        colorize=True,
    )

    # File handlers in production
    if settings.is_production:
        log_dir = Path("logs")
        log_dir.mkdir(exist_ok=True)
        _loguru_logger.add(
            str(log_dir / "{extra[service]}-error.log"),
            level="ERROR",
            rotation="5 MB",
            retention=5,
            format="{time:YYYY-MM-DD HH:mm:ss} [{extra[service]}] {level}: {message}",
        )
        _loguru_logger.add(
            str(log_dir / "{extra[service]}-combined.log"),
            rotation="10 MB",
            retention=5,
            format="{time:YYYY-MM-DD HH:mm:ss} [{extra[service]}] {level}: {message}",
        )


# Run setup on import
_setup_logger()


def create_logger(service: str = "app"):
    """Create a logger instance bound to a service name.

    Args:
        service: Service identifier ('auth', 'agent', 'server')

    Returns:
        A loguru logger bound with the service context.
    """
    return _loguru_logger.bind(service=service)


class RequestLoggingMiddleware(BaseHTTPMiddleware):
    """FastAPI middleware that logs each request with method, URL, status, and duration."""

    def __init__(self, app, logger=None):
        super().__init__(app)
        self.logger = logger or create_logger("server")

    async def dispatch(self, request: Request, call_next) -> Response:
        start = time.time()
        response = await call_next(request)
        duration_ms = round((time.time() - start) * 1000)

        status = response.status_code
        level = "error" if status >= 500 else "warning" if status >= 400 else "info"

        self.logger.log(
            level.upper(),
            f"{request.method} {request.url.path} {status} {duration_ms}ms",
        )
        return response
